src/debug.c

```diff
@@ -1975,8 +1975,9 @@ void bugReportEnd(int killViaSignal, int sig) {
     if (server.daemonize && server.supervised == 0 && server.pidfile) unlink(server.pidfile);
 
     if (!killViaSignal) {
+        /* To avoid issues with valgrind, we may wanna exit rahter than generate a signal */
         if (server.use_exit_on_panic)
-            exit(1);
+            _exit(1); /* Using _exit to bypass false leak reports by gcc ASAN */
```

Comment:
@oranagra Maybe define a `panic_exit()` function for these cases? Might be useful in other ways down the road.

Reply:
you mean that `panic_exit()` will also check the `use_exit_on_panic` flag and do `abort()`?
or just call `_exit()`?

i think i'd rather leave that for when we find the use for it rather than do it now..

but are you saying you're ok with using _exit here?
p.s. it means that asan leak detector is also not able to detect leaks when a slave exists (see `exitFromChild`), unlike valgrind which i think works on that case.

Reply:
@oranagra Also, AFAIK _exit() won't flush stdout and stderr. We may add manual calls for those before _exit(); 

```c
fflush(stderr);
fflush(stdout);
_exit(1);
```

Reply:
good point.
we normally don't use either of them, but there's are cases where we don't have a log file and instead log to stdout.
IIRC stderr is not cached.

maybe there could be other issues with not calling proper `exit()`, like some modules or libraries that use `atexit()`?

and the question remains, why is gcc ASAN broken, when clang ASAN is fine (it's the same code, right?).
and if we don't want to take the alternative of just disable the leak-check in gcc?
i.e. the benefit of what i did is that we keep the leak check for normal exit, and just give it up on assertions.

maybe we should use clang for the leak check we have in the PR CI (ci.yml)?

Reply:
> why is gcc ASAN broken, when clang ASAN is fine (it's the same code, right?).

They might be using different versions of ASAN. 

Btw I get same failure with Clang as well (Clang-12). Which version are you using?

Reply:
bionic gave me clang v6.0.0 and gcc v7.5.0
just confirmed again, clan doesn't warn about that leak.
but since yours is newer, i suppose that's the trend...

Reply:
@oranagra yes, I meant a single function (or macro) to deal with that - but not feeling strongly about it.

Reply:
I pulled your branch again, now I see that failure with gcc-9 but not with gcc-10 and clang-12. Either something is random or I somehow didn't compile with clang fully (I'm pretty sure I saw clang command printed on cmdline though).

Reply:
you mean pulled the branch (for the test), and reverted the `_exit` change... right?
so bottom line, we're gonna add fflush and merge it?

Reply:
> you mean pulled the branch (for the test), and reverted the _exit change... right?
Yes, right. 

> so bottom line, we're gonna add fflush and merge it?
Your point about `atexit()` might be valid. I feel like nobody should rely on this behavior but I'm not sure :) Other than that,  it should work properly AFAIK.